### ISSUE TITLE
Filter DataSet or DataView for Timeline

### ISSUE THREAD

**Issue by tengl** [almende/vis#339]
Hi,

I've been trying to set up a dynamic but simple filter on a Timeline. I want to hide or display items based on a property on each item, e.g. if I check a checkbox I want to display items where item.itemType = 1. If I uncheck the checkbox I want to hide items where item.itemType = 1. 

How can I accomplish that? I've tried to use DataView with filter but it doesn't evaluate the filter when the conditions are changed. I've tried to use timeline.setItems(new DataView(originalData, {filter: ... )) but the items are still there even if they should be filtered out. Timeline.redraw() doesn't help either.

Any ideas what I should do?

Thanks for a great library otherwise.

**Comment by AlexDM0** [comment]
Hi,

Over the last year a lot of feature requests have been made. We have just introduced our new website which has a list of the requested features. We have placed this request on that list. 

The list can be found here:
http://visjs.org/featureRequests.html

An explaination of the new system can be found here:
http://visjs.org/blog.html#New\ website\ for\ vis\.js\!

I would like to stress that this does not mean we abandon this request. Discussion here will continue if needed on this feature but we will close it to keep our Github issue page more of a bug-todo list.

Future feature requests will still be made here and then added to the website by us.

Regards,

Alex

**Comment by josdejong** [comment]
I've implemented a function `DataView.refresh()` to update filtered data, you can find in the `develop` branch of the project. I didn't call the function `update` to prevent confusion with the `update` function that `DataSet` has.

**Comment by daniel-pfeiffer** [comment]
I had till now missed out on DataView.  Now I thought I got it, but it doesn't seem so.  I have wrapped my timeline groups DataSet in a DataView and put a filter function on it.  Each group has a `visible` attribute which I successfully toggle by checkbox.  My filter just returns this attribute.  But not even if I do `myTimeline.groups.refresh()` does the timeline take note, and not even if I `redraw()` that.

Shouldn't using a DataView make it listen automatically?  Is there some example code how to do this right?  Also for items, which are more important to me.

**Comment by josdejong** [comment]
The DataView listens for changes in the DataSet (add, remove, update of items). It cannot know however when to update in case of a filter function of which the output depends on some external variable (in your case the value of a checkbox). In that case, you will have to call `DataView.refresh()` after the checkbox value changed. See docs on `refresh`:

http://visjs.org/docs/data/dataview.html#Methods

**Comment by daniel-pfeiffer** [comment]
Thanks Jos!  My confusion was actually about the data.  There seem to be three separate copies of it:
- the one I pass in, which gets ignored once read, but I was updating it
- the one returned by get(), changes to which don't survive
- and apparently the internal one, that needs an explicit call to update()

But for the DataView-options object that luckily doesn't seem to be the case.  It would be cool if it could be guaranteed to stay that way, because I add and remove the filter as needed.  I would guess that to be cheaper than running it on every event, when I know they'd all return true.

**Comment by josdejong** [comment]
Sorry I don't really understand what problem you try to explain. The DataSet indeed returns a clone of items when you call `get()`. Why is that a problem? Do you have a performance problem?

**Comment by daniel-pfeiffer** [comment]
Well all is fine now.  Initially I had put an AngularJS scope on members of what I passed in to the timeline.  Due to the unexpected copying, later changes were not being noticed.  Then I switched to the object returned by get, again to no avail.  Finally I figured the one and only way is to actively call update.  From the doc that had seemed like just one possible way.  You might need to clarify it is the only way.  Anyway, thanks!
